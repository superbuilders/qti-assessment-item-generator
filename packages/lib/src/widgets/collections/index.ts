import { allWidgetsCollection } from "@/widgets/collections/all"
import { fourthGradeMathCollection } from "@/widgets/collections/fourth-grade-math"
import { mathCoreCollection } from "@/widgets/collections/math-core"
import { scienceCollection } from "@/widgets/collections/science"
import { simpleVisualCollection } from "@/widgets/collections/simple-visual"
import { teksMath4Collection } from "@/widgets/collections/teks-math-4"

export const widgetCollections: {
	readonly all: typeof allWidgetsCollection
	readonly science: typeof scienceCollection
	readonly "simple-visual": typeof simpleVisualCollection
	readonly "fourth-grade-math": typeof fourthGradeMathCollection
	readonly "teks-math-4": typeof teksMath4Collection
	readonly "math-core": typeof mathCoreCollection
} = {
	all: allWidgetsCollection,
	science: scienceCollection,
	"simple-visual": simpleVisualCollection,
	"fourth-grade-math": fourthGradeMathCollection,
	"teks-math-4": teksMath4Collection,
	"math-core": mathCoreCollection
}

export type { WidgetCollection } from "@/widgets/collections/types"
