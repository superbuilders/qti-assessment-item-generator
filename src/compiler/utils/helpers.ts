import { typedSchemas } from "@/widgets/registry"

/**
 * Calculates the greatest common divisor (GCD) of two numbers using the Euclidean algorithm.
 * @param a The first number.
 * @param b The second number.
 * @returns The GCD of a and b.
 */
export function gcd(a: number, b: number): number {
	return b === 0 ? a : gcd(b, a % b)
}

/**
 * Determines if a fraction results in a terminating decimal.
 * A fraction terminates if and only if the prime factors of its simplified denominator are only 2s and 5s.
 * @param numerator The numerator of the fraction.
 * @param denominator The denominator of the fraction.
 * @returns True if the fraction terminates, false otherwise.
 */
export function isTerminatingFraction(
	numerator: number,
	denominator: number
): boolean {
	if (denominator === 0) {
		return false
	}
	// Simplify the fraction by dividing by the GCD.
	const commonDivisor = gcd(Math.abs(numerator), Math.abs(denominator))
	let simplifiedDen = Math.abs(denominator) / commonDivisor

	// The fraction terminates if the simplified denominator's prime factors are only 2 and 5.
	while (simplifiedDen % 2 === 0) {
		simplifiedDen /= 2
	}
	while (simplifiedDen % 5 === 0) {
		simplifiedDen /= 5
	}

	return simplifiedDen === 1
}

export function encodeDataUri(content: string): string {
	const base64 = btoa(unescape(encodeURIComponent(content)))
	const isSvg = content.trim().startsWith("<svg")
	return `${isSvg ? "data:image/svg+xml" : "data:text/html"};base64,${base64}`
}

// Helper function to check if a string is a valid widget type
export function isValidWidgetType(
	type: string
): type is keyof typeof typedSchemas {
	return type in typedSchemas
}

/**
 * Attempts to approximate simple multiplicative/divisive expressions that include π or pi
 * by replacing pi with the provided approximation (default 3.14) and evaluating the expression.
 *
 * Supported forms (case-insensitive for "pi"):
 * - "144*pi", "pi*144", "144pi" (implicit multiplication), "pi144" (rare, but treated as pi*144)
 * - Chains with '*' and '/' only, e.g. "3/2*pi", "pi/4", "6*pi/5", "(144) * (pi)"
 * Parentheses are allowed but only for grouping; no addition/subtraction is supported.
 *
 * Returns a normalized decimal string if evaluation succeeds, otherwise null.
 */
export function tryApproximatePiProduct(
	rawExpr: string,
	piApprox = 3.14
): string | null {
	if (typeof rawExpr !== "string" || rawExpr.trim() === "") return null

	// Quick check: must contain a pi token of some sort
	const containsPi = /\bpi\b/i.test(rawExpr) || rawExpr.includes("π")
	if (!containsPi) return null

	// Normalize: insert explicit * for implicit numeric-pi adjacency in common cases
	const expr = rawExpr
		// insert * between number and pi (e.g., 144pi -> 144*pi)
		.replace(/(\d)\s*(π|\bpi\b)/gi, "$1*$2")
		// and between pi and number (e.g., pi144 -> pi*144)
		.replace(/(π|\bpi\b)\s*(\d)/gi, "$1*$2")
		// remove whitespace
		.replace(/\s+/g, "")
		// drop parentheses (multiplicative-only evaluation does not require them)
		.replace(/[()]/g, "")
		// normalize 'pi' variants to numeric approximation
		.replace(/π/gi, String(piApprox))
		.replace(/\bpi\b/gi, String(piApprox))

	// After substitution, we only allow digits, '.', '*', '/' characters
	if (/[^0-9.*/]/.test(expr)) return null
	// Disallow addition/subtraction in this lightweight evaluator
	if (/[+-]/.test(expr)) return null
	// Must still contain the approximation value to ensure we actually used pi
	if (!expr.includes(String(piApprox))) return null

	// Split into tokens by * and / and evaluate left-to-right (* and / are same precedence)
	const parts: string[] = expr.split(/([*/])/)
	if (parts.length === 0) return null

	// Validate tokens
	for (let i = 0; i < parts.length; i++) {
		const t = parts[i]
		if (i % 2 === 0) {
			// operand
			if (t === "" || Number.isNaN(Number(t))) return null
		} else {
			// operator must be * or /
			if (t !== "*" && t !== "/") return null
		}
	}

	let acc = Number(parts[0])
	for (let i = 1; i < parts.length; i += 2) {
		const op = parts[i]
		const rhs = Number(parts[i + 1])
		if (op === "*") acc = acc * rhs
		else {
			if (rhs === 0) return null
			acc = acc / rhs
		}
	}

	// Normalize number to a compact string while preserving significant digits
	// Avoid scientific notation for typical classroom values
	const normalizeNumberString = () => {
		const maybeFixed = acc.toFixed(10) // limit noise
		// trim trailing zeros
		const trimmed = maybeFixed.replace(/\.0+$/, "").replace(/(\.\d*?)0+$/, "$1")
		return trimmed
	}
	const asString = normalizeNumberString()

	// Add a guard: must be a finite number
	if (!Number.isFinite(Number(asString))) return null
	return asString
}

/**
 * Normalizes a string part for use in a derived QTI identifier.
 * Converts to upper snake case and removes invalid characters.
 * @param part The string part to normalize.
 * @returns The normalized string.
 */
export function normalizeIdPart(part: string): string {
	return part.toUpperCase().replace(/[^A-Z0-9_]/g, "_")
}

/**
 * Derives a deterministic, QTI-safe combo identifier from a path of response parts.
 * @param pathParts An array of strings, where each string is a response-value pair (e.g., "RESPONSE_1_A").
 * @returns The fully-formed combo identifier (e.g., "FB__RESPONSE_1_A__RESPONSE_2_CORRECT").
 */
export function deriveComboIdentifier(pathParts: string[]): string {
	return `FB__${pathParts.join("__")}`
}
