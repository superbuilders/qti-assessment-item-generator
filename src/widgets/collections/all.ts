import type { WidgetCollection } from "@/widgets/collections/types"
import { allWidgetDefinitions } from "@/widgets/definitions"

function toReadonlyKeys<T extends Record<string, unknown>>(
	obj: T
): ReadonlyArray<keyof T & string> {
	const keys = Object.keys(obj)
	// Narrow via a runtime check that keys are strings
	return keys.filter((k): k is keyof T & string => typeof k === "string")
}

export const allWidgetsCollection: WidgetCollection<
	typeof allWidgetDefinitions,
	ReadonlyArray<keyof typeof allWidgetDefinitions & string>
> = {
	name: "all",
	widgets: allWidgetDefinitions,
	widgetTypeKeys: toReadonlyKeys(allWidgetDefinitions)
} as const
