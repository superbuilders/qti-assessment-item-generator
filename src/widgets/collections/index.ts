import { allWidgetsCollection } from "./all"
import { fourthGradeMathCollection } from "./fourth-grade-math"
import { mathCoreCollection } from "./math-core"
import { scienceCollection } from "./science"
import { simpleVisualCollection } from "./simple-visual"
import { teksMath4Collection } from "./teks-math-4"

export const widgetCollections: {
	readonly all: typeof allWidgetsCollection
	readonly science: typeof scienceCollection
	readonly "simple-visual": typeof simpleVisualCollection
	readonly "fourth-grade-math": typeof fourthGradeMathCollection
	readonly "teks-math-4": typeof teksMath4Collection
	readonly "math-core": typeof mathCoreCollection
} = {
	all: allWidgetsCollection,
	science: scienceCollection,
	"simple-visual": simpleVisualCollection,
	"fourth-grade-math": fourthGradeMathCollection,
	"teks-math-4": teksMath4Collection,
	"math-core": mathCoreCollection
}

export type { WidgetCollection } from "./types"
