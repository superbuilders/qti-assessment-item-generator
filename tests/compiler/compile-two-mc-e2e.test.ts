import { describe, expect, it } from "bun:test"
import { compile } from "../../src/compiler/compiler"
import type { AssessmentItem } from "../../src/core/item"
import { allWidgetsCollection } from "../../src/widgets/collections/all"
import { MINIMAL_INCORRECT_FEEDBACK } from "../helpers/feedback-fixtures"

describe("E2E compile: two MC with combo feedback", () => {
	it("produces valid QTI with response processing and feedback blocks", async () => {
		const item: AssessmentItem<[]> = {
			identifier: "ITEM_TWO_MC",
			title: "Two MC E2E",
			responseDeclarations: [
				{ identifier: "RESPONSE_1", cardinality: "single", baseType: "identifier", correct: "A" },
				{ identifier: "RESPONSE_2", cardinality: "single", baseType: "identifier", correct: "B" }
			],
			body: [
				{ type: "paragraph", content: [{ type: "text", content: "Question stem." }] },
				{ type: "interactionRef", interactionId: "int_1" },
				{ type: "interactionRef", interactionId: "int_2" }
			],
			widgets: null,
			interactions: {
				int_1: {
					type: "choiceInteraction",
					responseIdentifier: "RESPONSE_1",
					prompt: [{ type: "text", content: "Pick A or B" }],
					choices: [
						{ identifier: "A", content: [{ type: "paragraph", content: [{ type: "text", content: "A" }] }] },
						{ identifier: "B", content: [{ type: "paragraph", content: [{ type: "text", content: "B" }] }] }
					],
					shuffle: true,
					minChoices: 1,
					maxChoices: 1
				},
				int_2: {
					type: "choiceInteraction",
					responseIdentifier: "RESPONSE_2",
					prompt: [{ type: "text", content: "Pick A or B" }],
					choices: [
						{ identifier: "A", content: [{ type: "paragraph", content: [{ type: "text", content: "A" }] }] },
						{ identifier: "B", content: [{ type: "paragraph", content: [{ type: "text", content: "B" }] }] }
					],
					shuffle: true,
					minChoices: 1,
					maxChoices: 1
				}
			},
			feedbackPlan: {
				mode: "combo",
				dimensions: [
					{ responseIdentifier: "RESPONSE_1", kind: "enumerated", keys: ["A", "B"] },
					{ responseIdentifier: "RESPONSE_2", kind: "enumerated", keys: ["A", "B"] }
				],
				combinations: [
					{
						id: "FB__RESPONSE_1_A__RESPONSE_2_A",
						path: [
							{ responseIdentifier: "RESPONSE_1", key: "A" },
							{ responseIdentifier: "RESPONSE_2", key: "A" }
						]
					},
					{
						id: "FB__RESPONSE_1_A__RESPONSE_2_B",
						path: [
							{ responseIdentifier: "RESPONSE_1", key: "A" },
							{ responseIdentifier: "RESPONSE_2", key: "B" }
						]
					},
					{
						id: "FB__RESPONSE_1_B__RESPONSE_2_A",
						path: [
							{ responseIdentifier: "RESPONSE_1", key: "B" },
							{ responseIdentifier: "RESPONSE_2", key: "A" }
						]
					},
					{
						id: "FB__RESPONSE_1_B__RESPONSE_2_B",
						path: [
							{ responseIdentifier: "RESPONSE_1", key: "B" },
							{ responseIdentifier: "RESPONSE_2", key: "B" }
						]
					}
				]
			},
			feedback: {
				FEEDBACK__OVERALL: {
					RESPONSE_1: {
						A: {
							RESPONSE_2: {
								A: { content: MINIMAL_INCORRECT_FEEDBACK },
								B: { content: MINIMAL_INCORRECT_FEEDBACK }
							}
						},
						B: {
							RESPONSE_2: {
								A: { content: MINIMAL_INCORRECT_FEEDBACK },
								B: { content: MINIMAL_INCORRECT_FEEDBACK }
							}
						}
					}
				}
			}
		}

		const xml = await compile(item, allWidgetsCollection)

		expect(xml).toContain("<qti-assessment-item")
		expect(xml).toContain('identifier="ITEM_TWO_MC"')
		expect(xml).toContain("<qti-response-processing>")
		expect(xml).toContain(
			'<qti-feedback-block outcome-identifier="FEEDBACK__OVERALL" identifier="FB__RESPONSE_1_A__RESPONSE_2_A"'
		)
		expect(xml).toContain(
			'<qti-feedback-block outcome-identifier="FEEDBACK__OVERALL" identifier="FB__RESPONSE_1_A__RESPONSE_2_B"'
		)
		expect(xml).toContain(
			'<qti-feedback-block outcome-identifier="FEEDBACK__OVERALL" identifier="FB__RESPONSE_1_B__RESPONSE_2_A"'
		)
		expect(xml).toContain(
			'<qti-feedback-block outcome-identifier="FEEDBACK__OVERALL" identifier="FB__RESPONSE_1_B__RESPONSE_2_B"'
		)

		// Snapshot the full QTI for real renderer verification
		expect(xml).toMatchSnapshot()
	})
})
