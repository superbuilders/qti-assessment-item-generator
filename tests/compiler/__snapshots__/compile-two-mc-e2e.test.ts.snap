// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`E2E compile: two MC with combo feedback produces valid QTI with response processing and feedback blocks 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<qti-assessment-item
    xmlns="http://www.imsglobal.org/xsd/imsqtiasi_v3p0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://www.imsglobal.org/xsd/imsqtiasi_v3p0 https://purl.imsglobal.org/spec/qti/v3p0/schema/xsd/imsqti_asiv3p0p1_v1p0.xsd http://www.w3.org/1998/Math/MathML https://purl.imsglobal.org/spec/mathml/v3p0/schema/xsd/mathml3.xsd"
    identifier="ITEM_TWO_MC"
    title="Two MC E2E"
    time-dependent="false"
    xml:lang="en-US">
    <qti-response-declaration identifier="RESPONSE_1" cardinality="single" base-type="identifier">
        <qti-correct-response>
            <qti-value>A</qti-value>
        </qti-correct-response>
    </qti-response-declaration>
    <qti-response-declaration identifier="RESPONSE_2" cardinality="single" base-type="identifier">
        <qti-correct-response>
            <qti-value>B</qti-value>
        </qti-correct-response>
    </qti-response-declaration>
    <qti-outcome-declaration identifier="SCORE" cardinality="single" base-type="float">
        <qti-default-value><qti-value>0</qti-value></qti-default-value>
    </qti-outcome-declaration>
    <qti-outcome-declaration identifier="FEEDBACK__OVERALL" cardinality="single" base-type="identifier"/>
    <qti-item-body>
        <p>Question stem.</p>
        <div><qti-choice-interaction response-identifier="RESPONSE_1" shuffle="true" min-choices="1" max-choices="1">
            <qti-prompt>Pick A or B</qti-prompt>
            <qti-simple-choice identifier="A"><p>A</p></qti-simple-choice>
            <qti-simple-choice identifier="B"><p>B</p></qti-simple-choice>
        </qti-choice-interaction></div>
        <div><qti-choice-interaction response-identifier="RESPONSE_2" shuffle="true" min-choices="1" max-choices="1">
            <qti-prompt>Pick A or B</qti-prompt>
            <qti-simple-choice identifier="A"><p>A</p></qti-simple-choice>
            <qti-simple-choice identifier="B"><p>B</p></qti-simple-choice>
        </qti-choice-interaction></div>
        <qti-feedback-block outcome-identifier="FEEDBACK__OVERALL" identifier="FB__RESPONSE_1_A__RESPONSE_2_A" show-hide="show">
            <qti-content-body><div class="qti-feedback-preamble" data-correctness="incorrect" style="margin-bottom:16px;">
      <p style="margin:0; font-weight:700; color:#B91C1C;">Not quite! Try again.</p>
      <p style="margin:8px 0 0 0; color:#111827;">Your answer is incorrect.</p>
    </div><ol class="qti-steps" style="margin:16px 0; padding:0; list-style:none; counter-reset: qti-step;"><li class="qti-step" style="position:relative; margin:16px 0; padding:16px 16px 16px 56px; border:1px solid #E5E7EB; border-left:6px solid #2563EB; border-radius:12px; background:#F9FAFB; box-shadow:0 1px 2px rgba(0,0,0,0.04);">
        <span class="qti-step-index" style="position:absolute; left:16px; top:16px; width:28px; height:28px; border-radius:9999px; background:#2563EB; color:#FFFFFF; display:inline-flex; align-items:center; justify-content:center; font-weight:700; font-size:14px;">1</span>
        <p class="qti-step-title" style="margin:0 0 8px 0; font-weight:700; font-size:16px; color:#111827;">Identify Misconception</p>
        <p>Let's review the concept.</p>
      </li><li class="qti-step" style="position:relative; margin:16px 0; padding:16px 16px 16px 56px; border:1px solid #E5E7EB; border-left:6px solid #059669; border-radius:12px; background:#F9FAFB; box-shadow:0 1px 2px rgba(0,0,0,0.04);">
        <span class="qti-step-index" style="position:absolute; left:16px; top:16px; width:28px; height:28px; border-radius:9999px; background:#059669; color:#FFFFFF; display:inline-flex; align-items:center; justify-content:center; font-weight:700; font-size:14px;">2</span>
        <p class="qti-step-title" style="margin:0 0 8px 0; font-weight:700; font-size:16px; color:#111827;">Worked Example</p>
        <p>Here is how to solve it.</p>
      </li></ol></qti-content-body>
        </qti-feedback-block>
        <qti-feedback-block outcome-identifier="FEEDBACK__OVERALL" identifier="FB__RESPONSE_1_A__RESPONSE_2_B" show-hide="show">
            <qti-content-body><div class="qti-feedback-preamble" data-correctness="incorrect" style="margin-bottom:16px;">
      <p style="margin:0; font-weight:700; color:#B91C1C;">Not quite! Try again.</p>
      <p style="margin:8px 0 0 0; color:#111827;">Your answer is incorrect.</p>
    </div><ol class="qti-steps" style="margin:16px 0; padding:0; list-style:none; counter-reset: qti-step;"><li class="qti-step" style="position:relative; margin:16px 0; padding:16px 16px 16px 56px; border:1px solid #E5E7EB; border-left:6px solid #2563EB; border-radius:12px; background:#F9FAFB; box-shadow:0 1px 2px rgba(0,0,0,0.04);">
        <span class="qti-step-index" style="position:absolute; left:16px; top:16px; width:28px; height:28px; border-radius:9999px; background:#2563EB; color:#FFFFFF; display:inline-flex; align-items:center; justify-content:center; font-weight:700; font-size:14px;">1</span>
        <p class="qti-step-title" style="margin:0 0 8px 0; font-weight:700; font-size:16px; color:#111827;">Identify Misconception</p>
        <p>Let's review the concept.</p>
      </li><li class="qti-step" style="position:relative; margin:16px 0; padding:16px 16px 16px 56px; border:1px solid #E5E7EB; border-left:6px solid #059669; border-radius:12px; background:#F9FAFB; box-shadow:0 1px 2px rgba(0,0,0,0.04);">
        <span class="qti-step-index" style="position:absolute; left:16px; top:16px; width:28px; height:28px; border-radius:9999px; background:#059669; color:#FFFFFF; display:inline-flex; align-items:center; justify-content:center; font-weight:700; font-size:14px;">2</span>
        <p class="qti-step-title" style="margin:0 0 8px 0; font-weight:700; font-size:16px; color:#111827;">Worked Example</p>
        <p>Here is how to solve it.</p>
      </li></ol></qti-content-body>
        </qti-feedback-block>
        <qti-feedback-block outcome-identifier="FEEDBACK__OVERALL" identifier="FB__RESPONSE_1_B__RESPONSE_2_A" show-hide="show">
            <qti-content-body><div class="qti-feedback-preamble" data-correctness="incorrect" style="margin-bottom:16px;">
      <p style="margin:0; font-weight:700; color:#B91C1C;">Not quite! Try again.</p>
      <p style="margin:8px 0 0 0; color:#111827;">Your answer is incorrect.</p>
    </div><ol class="qti-steps" style="margin:16px 0; padding:0; list-style:none; counter-reset: qti-step;"><li class="qti-step" style="position:relative; margin:16px 0; padding:16px 16px 16px 56px; border:1px solid #E5E7EB; border-left:6px solid #2563EB; border-radius:12px; background:#F9FAFB; box-shadow:0 1px 2px rgba(0,0,0,0.04);">
        <span class="qti-step-index" style="position:absolute; left:16px; top:16px; width:28px; height:28px; border-radius:9999px; background:#2563EB; color:#FFFFFF; display:inline-flex; align-items:center; justify-content:center; font-weight:700; font-size:14px;">1</span>
        <p class="qti-step-title" style="margin:0 0 8px 0; font-weight:700; font-size:16px; color:#111827;">Identify Misconception</p>
        <p>Let's review the concept.</p>
      </li><li class="qti-step" style="position:relative; margin:16px 0; padding:16px 16px 16px 56px; border:1px solid #E5E7EB; border-left:6px solid #059669; border-radius:12px; background:#F9FAFB; box-shadow:0 1px 2px rgba(0,0,0,0.04);">
        <span class="qti-step-index" style="position:absolute; left:16px; top:16px; width:28px; height:28px; border-radius:9999px; background:#059669; color:#FFFFFF; display:inline-flex; align-items:center; justify-content:center; font-weight:700; font-size:14px;">2</span>
        <p class="qti-step-title" style="margin:0 0 8px 0; font-weight:700; font-size:16px; color:#111827;">Worked Example</p>
        <p>Here is how to solve it.</p>
      </li></ol></qti-content-body>
        </qti-feedback-block>
        <qti-feedback-block outcome-identifier="FEEDBACK__OVERALL" identifier="FB__RESPONSE_1_B__RESPONSE_2_B" show-hide="show">
            <qti-content-body><div class="qti-feedback-preamble" data-correctness="incorrect" style="margin-bottom:16px;">
      <p style="margin:0; font-weight:700; color:#B91C1C;">Not quite! Try again.</p>
      <p style="margin:8px 0 0 0; color:#111827;">Your answer is incorrect.</p>
    </div><ol class="qti-steps" style="margin:16px 0; padding:0; list-style:none; counter-reset: qti-step;"><li class="qti-step" style="position:relative; margin:16px 0; padding:16px 16px 16px 56px; border:1px solid #E5E7EB; border-left:6px solid #2563EB; border-radius:12px; background:#F9FAFB; box-shadow:0 1px 2px rgba(0,0,0,0.04);">
        <span class="qti-step-index" style="position:absolute; left:16px; top:16px; width:28px; height:28px; border-radius:9999px; background:#2563EB; color:#FFFFFF; display:inline-flex; align-items:center; justify-content:center; font-weight:700; font-size:14px;">1</span>
        <p class="qti-step-title" style="margin:0 0 8px 0; font-weight:700; font-size:16px; color:#111827;">Identify Misconception</p>
        <p>Let's review the concept.</p>
      </li><li class="qti-step" style="position:relative; margin:16px 0; padding:16px 16px 16px 56px; border:1px solid #E5E7EB; border-left:6px solid #059669; border-radius:12px; background:#F9FAFB; box-shadow:0 1px 2px rgba(0,0,0,0.04);">
        <span class="qti-step-index" style="position:absolute; left:16px; top:16px; width:28px; height:28px; border-radius:9999px; background:#059669; color:#FFFFFF; display:inline-flex; align-items:center; justify-content:center; font-weight:700; font-size:14px;">2</span>
        <p class="qti-step-title" style="margin:0 0 8px 0; font-weight:700; font-size:16px; color:#111827;">Worked Example</p>
        <p>Here is how to solve it.</p>
      </li></ol></qti-content-body>
        </qti-feedback-block>
    </qti-item-body>
    <qti-response-processing>
        
    <qti-response-condition>
        <qti-response-if>
            <qti-match>
                <qti-variable identifier="RESPONSE_1"/>
                <qti-base-value base-type="identifier">A</qti-base-value>
            </qti-match>
    <qti-response-condition>
        <qti-response-if>
            <qti-match>
                <qti-variable identifier="RESPONSE_2"/>
                <qti-base-value base-type="identifier">A</qti-base-value>
            </qti-match>
            <qti-set-outcome-value identifier="FEEDBACK__OVERALL">
                <qti-base-value base-type="identifier">FB__RESPONSE_1_A__RESPONSE_2_A</qti-base-value>
            </qti-set-outcome-value>
        </qti-response-if>
        <qti-response-else-if>
            <qti-match>
                <qti-variable identifier="RESPONSE_2"/>
                <qti-base-value base-type="identifier">B</qti-base-value>
            </qti-match>
            <qti-set-outcome-value identifier="FEEDBACK__OVERALL">
                <qti-base-value base-type="identifier">FB__RESPONSE_1_A__RESPONSE_2_B</qti-base-value>
            </qti-set-outcome-value>
        </qti-response-else-if>
    </qti-response-condition>
        </qti-response-if>
        <qti-response-else-if>
            <qti-match>
                <qti-variable identifier="RESPONSE_1"/>
                <qti-base-value base-type="identifier">B</qti-base-value>
            </qti-match>
    <qti-response-condition>
        <qti-response-if>
            <qti-match>
                <qti-variable identifier="RESPONSE_2"/>
                <qti-base-value base-type="identifier">A</qti-base-value>
            </qti-match>
            <qti-set-outcome-value identifier="FEEDBACK__OVERALL">
                <qti-base-value base-type="identifier">FB__RESPONSE_1_B__RESPONSE_2_A</qti-base-value>
            </qti-set-outcome-value>
        </qti-response-if>
        <qti-response-else-if>
            <qti-match>
                <qti-variable identifier="RESPONSE_2"/>
                <qti-base-value base-type="identifier">B</qti-base-value>
            </qti-match>
            <qti-set-outcome-value identifier="FEEDBACK__OVERALL">
                <qti-base-value base-type="identifier">FB__RESPONSE_1_B__RESPONSE_2_B</qti-base-value>
            </qti-set-outcome-value>
        </qti-response-else-if>
    </qti-response-condition>
        </qti-response-else-if>
    </qti-response-condition>
    <qti-response-condition>
        <qti-response-if>
            <qti-and>
                <qti-match><qti-variable identifier="RESPONSE_1"/><qti-correct identifier="RESPONSE_1"/></qti-match>
                        <qti-match><qti-variable identifier="RESPONSE_2"/><qti-correct identifier="RESPONSE_2"/></qti-match>
            </qti-and>
            <qti-set-outcome-value identifier="SCORE"><qti-base-value base-type="float">1.0</qti-base-value></qti-set-outcome-value>
        </qti-response-if>
        <qti-response-else>
            <qti-set-outcome-value identifier="SCORE"><qti-base-value base-type="float">0.0</qti-base-value></qti-set-outcome-value>
        </qti-response-else>
    </qti-response-condition>
    </qti-response-processing>
</qti-assessment-item>"
`;
